# <font color='blue'>Data Science Academy - Python Fundamentos - Capítulo 9</font>

## Download: http://github.com/dsacademybr

## Exercício: Análise Exploratória de Dados com Python

Neste exercício, você vai realizar uma análise exploratória em um dos mais famosos datasets para Machine Learning, o dataset iris com informações sobre 3 tipos de plantas. Esse dataset é comumente usado em problemas de Machine Learning de classificação, quando nosso objetivo é prever a classe dos dados. No caso deste dataset, prever a categoria de uma planta a partir de medidas da planta (sepal e petal).

Em cada célula, você encontra a tarefa a ser realizada. Faça todo o exercício e depois compare com a solução proposta.

Dataset (já disponível com o Scikit-Learn): https://archive.ics.uci.edu/ml/datasets/iris

In [1]:
# Imports
import time
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from sklearn.datasets import load_iris
%matplotlib inline

fontsize = 14
ticklabelsize = 14

In [2]:
# Carregando o dataset
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
print(len(df))
df.head()

150


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

## Extração e Transformação de Dados

In [5]:
# Imprima os valores numéricos da Variável target (o que queremos prever), 
# uma de 3 possíveis categorias de plantas: setosa, versicolor ou virginica
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [13]:
# Imprima os valores numéricos da Variável target (o que queremos prever), 
# uma de 3 possíveis categorias de plantas: 0, 1 ou 2
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [18]:
# Adicione ao dataset uma nova coluna com os nomes das espécies, pois é isso que vamos tentar prever (variável target)
df['species'] = pd.Categorical.from_codes(iris.target, iris.target_names)


<bound method NDFrame.head of      sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
5                  5.4               3.9                1.7               0.4   
6                  4.6               3.4                1.4               0.3   
7                  5.0               3.4                1.5               0.2   
8                  4.4               2.9                1.4               0.2   
9                  4.9               3.1                1.5               0.1   
10                 5.4               3.7                1.5               0.2  

In [23]:
# Inclua no dataset uma coluna com os valores numéricos da variável target
df['target'] = iris.target
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
5                  5.4               3.9                1.7               0.4   
6                  4.6               3.4                1.4               0.3   
7                  5.0               3.4                1.5               0.2   
8                  4.4               2.9                1.4               0.2   
9                  4.9               3.1                1.5               0.1   
10                 5.4               3.7                1.5               0.2   
11                 4.8               3.4                1.6               0.2   
12                 4.8               3.0                1.4               0.1   
13                 4.3               3.0                1.1               0.1   
14                 5.8               4.0                1.2               0.2   
15                 5.7               4.4                1.5               0.4   
16                 5.4               3.9                1.3               0.4   
17                 5.1               3.5                1.4               0.3   
18                 5.7               3.8                1.7               0.3   
19                 5.1               3.8                1.5               0.3   
20                 5.4               3.4                1.7               0.2   
21                 5.1               3.7                1.5               0.4   
22                 4.6               3.6                1.0               0.2   
23                 5.1               3.3                1.7               0.5   
24                 4.8               3.4                1.9               0.2   
25                 5.0               3.0                1.6               0.2   
26                 5.0               3.4                1.6               0.4   
27                 5.2               3.5                1.5               0.2   
28                 5.2               3.4                1.4               0.2   
29                 4.7               3.2                1.6               0.2   
..                 ...               ...                ...               ...   
120                6.9               3.2                5.7               2.3   
121                5.6               2.8                4.9               2.0   
122                7.7               2.8                6.7               2.0   
123                6.3               2.7                4.9               1.8   
124                6.7               3.3                5.7               2.1   
125                7.2               3.2                6.0               1.8   
126                6.2               2.8                4.8               1.8   
127                6.1               3.0                4.9               1.8   
128                6.4               2.8                5.6               2.1   
129                7.2               3.0                5.8               1.6   
130                7.4               2.8                6.1               1.9   
131                7.9               3.8                6.4               2.0   
132                6.4               2.8                5.6               2.2   
133                6.3               2.8                5.1               1.5   
134                6.1               2.6                5.6               1.4   
135                7.7               3.0                6.1               2.3   
136                6.3               3.4                5.6               2.4   
137                6.4              

In [25]:
# Extraia as features (atributos) do dataset e imprima 
list(df)

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)',
 'species',
 'target']

In [8]:
# Calcule a média de cada feature para as 3 classes


## Exploração de Dados

In [32]:
# Imprima uma Transposta do dataset (transforme linhas e colunas e colunas em linhas)
df.T

0       1       2       3       4       5       6    \
sepal length (cm)     5.1     4.9     4.7     4.6       5     5.4     4.6   
sepal width (cm)      3.5       3     3.2     3.1     3.6     3.9     3.4   
petal length (cm)     1.4     1.4     1.3     1.5     1.4     1.7     1.4   
petal width (cm)      0.2     0.2     0.2     0.2     0.2     0.4     0.3   
species            setosa  setosa  setosa  setosa  setosa  setosa  setosa   
target                  0       0       0       0       0       0       0   

                      7       8       9      ...            140        141  \
sepal length (cm)       5     4.4     4.9    ...            6.7        6.9   
sepal width (cm)      3.4     2.9     3.1    ...            3.1        3.1   
petal length (cm)     1.5     1.4     1.5    ...            5.6        5.1   
petal width (cm)      0.2     0.2     0.1    ...            2.4        2.3   
species            setosa  setosa  setosa    ...      virginica  virginica   
target                  0       0       0    ...              2          2   

                         142        143        144        145        146  \
sepal length (cm)        5.8        6.8        6.7        6.7        6.3   
sepal width (cm)         2.7        3.2        3.3          3        2.5   
petal length (cm)        5.1        5.9        5.7        5.2          5   
petal width (cm)         1.9        2.3        2.5        2.3        1.9   
species            virginica  virginica  virginica  virginica  virginica   
target                     2          2          2          2          2   

                         147        148        149  
sepal length (cm)        6.5        6.2        5.9  
sepal width (cm)           3        3.4          3  
petal length (cm)        5.2        5.4        5.1  
petal width (cm)           2        2.3        1.8  
species            virginica  virginica  virginica  
target                     2          2          2  

[6 rows x 150 columns]

In [37]:
# Utilize a função Info do dataset para obter um resumo sobre o dataset 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
species              150 non-null category
target               150 non-null int64
dtypes: category(1), float64(4), int64(1)
memory usage: 6.2 KB


In [34]:
# Faça um resumo estatístico do dataset
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.054000           3.758667   
std             0.828066          0.433594           1.764420   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.198667    1.000000  
std            0.763161    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [36]:
# Verifique se existem valores nulos no dataset
df.isnull()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                False             False              False             False   
1                False             False              False             False   
2                False             False              False             False   
3                False             False              False             False   
4                False             False              False             False   
5                False             False              False             False   
6                False             False              False             False   
7                False             False              False             False   
8                False             False              False             False   
9                False             False              False             False   
10               False             False              False             False   
11               False             False              False             False   
12               False             False              False             False   
13               False             False              False             False   
14               False             False              False             False   
15               False             False              False             False   
16               False             False              False             False   
17               False             False              False             False   
18               False             False              False             False   
19               False             False              False             False   
20               False             False              False             False   
21               False             False              False             False   
22               False             False              False             False   
23               False             False              False             False   
24               False             False              False             False   
25               False             False              False             False   
26               False             False              False             False   
27               False             False              False             False   
28               False             False              False             False   
29               False             False              False             False   
..                 ...               ...                ...               ...   
120              False             False              False             False   
121              False             False              False             False   
122              False             False              False             False   
123              False             False              False             False   
124              False             False              False             False   
125              False             False              False             False   
126              False             False              False             False   
127              False             False              False             False   
128              False             False              False             False   
129              False             False              False             False   
130              False             False              False             False   
131              False             False              False             False   
132              False             False              False             False   
133              False             False              False             False   
134              False             False              False             False   
135              False             False              False             False   
136              False             False              False             False   
137              False             F

In [42]:
# Faça uma contagem de valores de sepal length
df.count()

sepal length (cm)    150
sepal width (cm)     150
petal length (cm)    150
petal width (cm)     150
species              150
target               150
dtype: int64

## Plot

In [14]:
# Crie um Histograma de sepal length


In [15]:
# Crie um Gráficos de Dispersão (scatter Plot) da variável sepal length versus número da linha, 
# colorido por marcadores da variável target


In [16]:
# Crie um Scatter Plot de 2 Features (atributos)


In [17]:
# Crie um Scatter Matrix das Features (atributos)


In [18]:
# Crie um Histograma de todas as features


Conheça a Formação Cientista de Dados, um programa completo, 100% online e 100% em português, com 340 horas, mais de 1.200 aulas em vídeos e 26 projetos, que vão ajudá-lo a se tornar um dos profissionais mais cobiçados do mercado de análise de dados. Clique no link abaixo, faça sua inscrição, comece hoje mesmo e aumente sua empregabilidade:

https://www.datascienceacademy.com.br/pages/formacao-cientista-de-dados

# Fim

### Obrigado - Data Science Academy - <a href="http://facebook.com/dsacademybr">facebook.com/dsacademybr</a>